### PR TITLE
Up to question 32

### PR DESCRIPTION
**Pull request by avnerlavi**

### CODE REVIEW

File: trainer.py [hunk 1]
```diff
@@ -112,9 +111,8 @@ def evaluate_model_on_dataloader(
                 loss = self.criterion(predictions, targets)
 
                 correct_labeled_samples += (predictions.argmax(1) == targets).type(torch.int).sum().item()
-                nof_samples += len(inputs)
                 total_loss += loss.item()
-                avg_loss = total_loss / nof_samples
+                avg_loss = total_loss / (batch_idx + 1)
                 accuracy = correct_labeled_samples * 100 / nof_samples
```
Comment (lev1ayz) [inline trainer.py:116]: should be:
accuracy = correct_labeled_samples * 100 / ((batch_idx + 1) * self.batch_size))

you didnt go over all the samples yet so its unfair to divide by the entire set

File: trainer.py [hunk 2]
```diff
@@ -62,16 +62,15 @@ def train_one_epoch(self) -> tuple[float, float]:
...
+            self.optimizer.zero_grad()
             loss.backward()
             self.optimizer.step()
 
             correct_labeled_samples += (predictions.argmax(1) == targets).type(torch.int).sum().item()
-            nof_samples += len(inputs)
             total_loss += loss.item()
-            avg_loss = total_loss / nof_samples
+            avg_loss = total_loss / (batch_idx + 1)
             accuracy = correct_labeled_samples * 100 / nof_samples
```
Comment (lev1ayz) [inline trainer.py:74]: should be:
accuracy = correct_labeled_samples * 100 / ((batch_idx + 1) * self.batch_size))

you didnt go over all the samples yet so its unfair to divide by the entire set

File: trainer.py [hunk 3]
```diff
@@ -51,7 +51,7 @@ def train_one_epoch(self) -> tuple[float, float]:
         total_loss = 0
         avg_loss = 0
         accuracy = 0
-        nof_samples = 0
+        nof_samples = len(self.train_dataset)
```
Comment (lev1ayz) [inline trainer.py]: change back to 0, and update during batch iteration
nof_samples += self.batch_size  (or something like that)

File: trainer.py [hunk 4]
```diff
@@ -100,7 +99,7 @@ def evaluate_model_on_dataloader(
         total_loss = 0
         avg_loss = 0
         accuracy = 0
-        nof_samples = 0
+        nof_samples = len(dataset)
```
Comment (lev1ayz) [inline trainer.py]: hange back to 0, and update during batch iteration
nof_samples += self.batch_size (or something like that)

**Review by lev1ayz — APPROVED** [review]
